libbeat/docs/upgrading.asciidoc

```diff
@@ -142,8 +142,18 @@ migration.6_to_7.enabled: true
 ----
 
 The field aliases enable you to use 6.x dashboards and visualizations
-with indices created by {beats} 7.0 or later. The aliases do *not* work
-with saved searches or with API calls that manipulate documents directly.
+with indices created by {beats} 7.0 or later. However, there are a couple of
+exceptions:
+
+* The aliases do *not* work with saved searches or with API calls that
+manipulate documents directly.
+* The aliases do *not* resolve type mismatches caused by field type changes
```

Comment:
Second point is confusing in how it's stated, IMO. I'm not sure people would expect aliases to help if the field is also changing type.

Starting from another point of view, we can even express this part positive and part caveat:

> Fields that are both moving with an alias, and whose type is changing will show up as a normal conflict in Kibana. This also means Kibana will not perform aggregations on these fields, like a normal type change.

I'm not super happy about the wording I'm proposing, but I think approaching it like this conveys the situation more clearly

Reply:
@webmat I think the statement is still potentially confusing. How about we simplify it by saying something like:

"Some of the fields renamed in 7.0 also have type changes. Kibana will identify the type conflicts and no longer perform aggregations on those fields."

It would also be nice to tell users (perhaps in the troubleshooting section) what actions to take to fix the problem. Still not quite sure what we should advise here.

Reply:
Yeah that sentence of mine had too much packed into it, indeed. And it occurs to me that perhaps we should plainly list all 3 situations.

- Some fields are changing type, your aggregations and visualizations will not work on those until the conflict is resolved
- Some fields are changing name and keeping the same type, the compatibility layer will create field aliases that ensure visualizations and aggregations on the old field names work on old and new data. The new dashboards only work on the new fields, and will therefore only work with the new data.
- Some fields are both being renamed and changing type. A field alias will be created for them, but will only be useful for Elasticsearch API queries in some cases. Kibana will not let you query them until the conflict is resolved.

Now on the "how to resolve conflicts": I would have thought we already had a page with some ideas & their caveats, after this many major versions, but it seems like that's not the case :-/

I think we should address this for the 6 to 7 migration, because there's more changes now than ever. I'm happy to contribute ideas there, but I think it's going to be a mini-project in itself, so maybe we can close the loop here simply, and start a page for this after? It would probably deserve its own documentation page, because there's many different strategies to resolve these conflicts.

WDYT?

cc @urso @ruflin do you know if we have a page or a blog post that goes over the different strategies for people to address conflicts during migrations like this?

Reply:
@webmat 

> I would have thought we already had a page with some ideas & their caveats

You're right. This isn't a Beats or ECS thing. I searched for a page about resolving mapping conflicts, but didn't find anything relevant. Ideally we would link to a page that explains how to resolve mapping conflicts. Linking to blog posts can be tricky because blogs get stale over time. 

> A field alias will be created for them, but will only be useful for Elasticsearch API queries in some cases.

This sort of begs the question: what do we mean by "in some cases."

Reply:
Pfff, it's impossible to let a fast one slip by, with you 😆 

When a field type is different in indices you're querying together, it's sometimes possible to run an aggregation over it, **with Elasticsearch directly**.

By "sometimes" I mean depending on what datatypes are at play (I haven't tried all datatype conflicts, this is just one that works and one that doesn't):

- If you have `keyword` and `ip`, a bucket aggregation will work.
- If you have `keyword` and `long`, a bucket aggregation will fail with something like this:

```JSON
# GET conflict-*/_search
{
  "error": {
    "root_cause": [],
    "type": "search_phase_execution_exception",
    "reason": "",
    "phase": "fetch",
    "grouped": true,
    "failed_shards": [],
    "caused_by": {
      "type": "aggregation_execution_exception",
      "reason": "Merging/Reducing the aggregations failed when computing the aggregation [codes] because the field you gave in the aggregation query existed as two different types in two different indices"
    }
  },
  "status": 503
}
```

Kibana, on the other hand, will simply not try aggregations on fields that are known to be in conflict in the index pattern.

So looping back to my earlier comment, aliases we're creating for fields that change name and also have a datatype change will only be useable in Elasticsearch API calls, not Kibana aggs/viz at this time.

If you want to experiment with this:

```JSON
# Create conflicting indices
PUT conflict-legacy
{ "mappings" : { "properties": {
  "address" : { "type" : "keyword" },
  "code" : { "type" : "keyword" }
}}}
PUT conflict-legacy/_doc/1
{ "address": "10.42.42.42", "code": "200" }

PUT conflict-new
{ "mappings" : { "properties": {
  "address" : { "type" : "ip" },
  "code" : { "type" : "long" }
}}}
PUT conflict-new/_doc/1
{ "address": "10.42.42.42", "code": 200 }

# Successful aggregation over "address"
GET conflict-*/_search
{ "size": 0, "aggs" : {
    "addresses" : { "terms" : { "field" : "address" }
}}}

# Failed aggregation (error above)
GET conflict-*/_search
{ "size": 0, "aggs" : {
    "codes" : { "terms" : { "field" : "code" }
}}}
```

Reply:
Once you've created these two small indices, you can of course try the same aggregations by creating an index pattern over `conflict-*`, then trying simple bucket aggregations with a table visualization.

In order to do so, however, you'll have to create the index pattern after creating one index, but before creating the other. If the field is already in conflict, you won't even be able to create the visualization at all, iirc the conflicted field will not appear in the field selector.

Reply:
@webmat Thanks for the detailed response. I've tried to incorporate your feedback into the the latest commit. I was worried that the content was getting too dense, so I added highlighting. I worry that users will not read all the words...there are so many caveats. Maybe it's OK for a first pass, though, and we can refine it later. I'll see if Karen can put her editor brain to work on it.